src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/Metadata/FormFileBindingSourceMetadataProvider.cs

```diff
@@ -0,0 +1,45 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Reflection;
+using Microsoft.AspNetCore.Http;
+
+namespace Microsoft.AspNetCore.Mvc.ModelBinding.Metadata
+{
+    public class FormFileBindingSourceMetadataProvider : IBindingMetadataProvider
+    {
+        public Type Type { get; }
+
+        /// <summary>
+        /// Creates a new <see cref="FormFileBindingSourceMetadataProvider"/> for the given <paramref name="type"/>. 
+        /// </summary>
+        /// <param name="type">
+        /// The <see cref="Type"/>. The provider sets <see cref="BindingSource"/> to <see cref="BindingSource.FormFile"/>
+        /// for properties of <see cref="Type"/> <see cref="IFormFile"/> and <see cref="IFormCollection"/>.
+        /// </param>
+        public FormFileBindingSourceMetadataProvider(Type type)
+        {
+            if (type == null)
+            {
+                throw new ArgumentNullException(nameof(type));
+            }
+
+            Type = type;
+        }
+
+        /// <inheritdoc />
+        public void CreateBindingMetadata(BindingMetadataProviderContext context)
+        {
+            if (context == null)
+            {
+                throw new ArgumentNullException(nameof(context));
+            }
+
+            if (Type.IsAssignableFrom(context.Key.ModelType))
```

Comment:
This doesn't do what the docs say 😆 this will set the source for the given type and **anything assignable** to the given type.